k2/csrc/fsa.cu

```diff
@@ -9,6 +9,134 @@
 
 namespace k2 {
 
+
+
+int32_t GetFsaVecBasicProperties(FsaVec &fsa_vec) {
+  if (fsa_vec.NumAxes() != 3) {
+    K2_LOG(FATAL) << "FsaVec has wrong number of axes: " << fsa_vec.NumAxes()
+                  << ", expected 3.";
+  }
+  Context c = fsa_vec.Context();
+  const int32_t *row_ids1_data = fsa_vec.RowIds1().Data(),
+      *row_splits1_data = fsa_vec.RowSplits1().Data(),
+      *row_ids2_data = fsa_vec.RowIds2().Data(),
+      *row_splits2_data = fsa_vec.RowSplits().Data();
+  Arc *arcs_data = fsa_vec.values.Data();
+
+  int32_t num_arcs = fsa_vec.values.Dim();
+
+  // `neg_` means negated.  It's more convenient to do it this way.
+  Array1<int32_t> neg_properties(c, num_arcs);
+  int32_t num_states = fsa_vec.RowIds1().Dim(),
+      num_fsas = fsa_vec.shape.TotSize(0);
+
+  // `reachable[idx01]` will be true if the state with index idx01 has an arc
+  // entering it or is state 0 of its FSA, not counting self-loops; it's a
+  // looser condition than being 'accessible' in FSA terminlogy, simply meaning
+  // it's reachable from some state (which might not itself be reachable).
+  //
+  // reachable[num_states + idx01] will be true if the state with index idx01 is
+  // the final-state of its FSA (i.e. last-numbered) or has at least one arc
+  // leaving it, not counting self-loops.  Again, it's a looser condition than
+  // being 'co-accessible' in FSA terminology.
+  Array1<char> reachable(c, num_states * 2 + 1, 0);
+  Array1<char> flag = reachable.Range(num_states * 1, 1);
+  Array1<char> co_reachable = reachable.Range(num_states, num_states);
+  reachable = reachable.Range(0, num_states);
+  int32_t *neg_properties_data = properties.Data();
+  char *reachable_data = reachable.Data();  // access co_reachable via this.
+
+  auto lambda_get_properties = [=] __host__ __device__ (int32_t idx012) -> void {
+      Arc arc = arcs_data[idx012];
+      Arc prev_arc;
+      if (idx012 > 0) prev_arc = arcs_data[idx012 - 1];
+      int32_t idx01 = row_ids2_data[idx012],
+          idx01x = row_splits2_data[idx012],
```

Comment:
Is this a typo or an error?

Should it be
```cpp
idx01x = row_splits2_data[idx01];
```
?

Reply:
yes thanks, you are right.  (This is one of the benefits of this naming system.. would have been impossible to spot this error without it.)